src/java/org/apache/cassandra/db/DeletionTime.java

```diff
@@ -195,42 +195,103 @@ public static Serializer getSerializer(Version version)
+                if ((flags & 0xFF) != IS_LIVE_DELETION)
+                    throw new IOException("Corrupted sstable. Invalid flags found deserializing DeletionTime");
+                return LIVE;
+            }
+            else
+            {
+                // Read the remaining 7 bytes
+                int bytes1 = in.readByte();
+                int bytes2 = in.readShort();
+                int bytes4 = in.readInt();
+
+                long mfda = readBytesToMFDA(flags, bytes1, bytes2, bytes4);
+                int localDeletionTimeUnsignedInteger = in.readInt();
+
+                return new DeletionTime(mfda, localDeletionTimeUnsignedInteger);
+            }
         }
 
-        public DeletionTime deserialize(ByteBuffer buf, int offset)
+        public DeletionTime deserialize(ByteBuffer buf, int offset) throws IOException
         {
-            int localDeletionTimeUnsignedInteger = buf.getInt(offset);
-            long mfda = buf.getLong(offset + 4);
-            return mfda == Long.MIN_VALUE && localDeletionTimeUnsignedInteger == Cell.NO_DELETION_TIME_UNSIGNED_INTEGER
-                   ? LIVE
-                   : new DeletionTime(mfda, localDeletionTimeUnsignedInteger);
+            int flags = buf.get(offset);
+            if ((flags & IS_LIVE_DELETION) != 0)
+            {
+                if ((flags & 0xFF) != IS_LIVE_DELETION)
+                    throw new IOException("Corrupted sstable. Invalid flags found deserializing DeletionTime");
+                return LIVE;
+            }
+            else
+            {
+                long mfda = buf.getLong(offset);
+                int localDeletionTimeUnsignedInteger = buf.getInt(offset + TypeSizes.LONG_SIZE);
+
+                return new DeletionTime(mfda, localDeletionTimeUnsignedInteger);
+            }
         }
 
         public void skip(DataInputPlus in) throws IOException
         {
-            in.skipBytesFully(4 + 8);
+            int flags = in.readByte();
+            if ((flags & IS_LIVE_DELETION) != 0)
+            {
+                if ((flags & 0xFF) != IS_LIVE_DELETION)
+                    throw new IOException("Corrupted sstable. Invalid flags found deserializing DeletionTime");
+                // We read the flags already and there's nothing left to skip over
+                return;
+            }
+            else
+                // We read the flags already but there is mfda and ldt to skip over still
+                in.skipBytesFully(TypeSizes.LONG_SIZE - 1 + TypeSizes.INT_SIZE);
         }
 
         public long serializedSize(DeletionTime delTime)
         {
-            return TypeSizes.sizeof(Integer.MAX_VALUE)
-                   + TypeSizes.sizeof(delTime.markedForDeleteAt());
+            if (delTime == LIVE)
+                // Just the flags
+                return 1;
+            else
+                // 1 for the flags, 7 for mfda and 4 for ldt
+                return TypeSizes.INT_SIZE
+                       + TypeSizes.LONG_SIZE;
```

Comment:
OCD-style nit: The order is different in the comment than in the actual return. The comment first mentions the long and int and the return is reversed. My initial reaction was "how do I fit 1 and 7 bytes in an int" ;)

Reply:
Will add on commit